cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -1602,7 +1602,17 @@ pub(crate) fn emit(
 
         Inst::Args { .. } => {}
 
-        Inst::Ret { .. } => sink.put1(0xC3),
+        Inst::Ret {
+            stack_bytes_to_pop: 0,
+            ..
+        } => sink.put1(0xC3),
+
+        Inst::Ret {
+            stack_bytes_to_pop, ..
+        } => {
+            sink.put1(0xC2);
+            sink.put2(u16::try_from(*stack_bytes_to_pop).unwrap());
```

Comment:
Less commonly used instructions from the 16bit x86 era can be slower than writing multiple simpler instructions. Does this also apply to `retn` or is it just as fast as `add rsp, n` + `ret`?

Reply:
According to [Agner Fog's instruction tables](https://www.agner.org/optimize/instruction_tables.pdf), and picking a random semi-recent microarchitecture (Skylake), `ret` and `ret imm` are both 2 µops; the latter has a throughput of 0.5 per cycle, the former 1 per cycle. So it looks like there's a tiny penalty but it's no worse than rearranging the stack manually. (And keep in mind that the popped data is *above* the return address, so one would need to load that, store it higher up, move rsp, and `ret`; the store alone in that sequence would kill the perf advantage.)

Reply:
`add rsp, n; ret` isn't equivalent because the return address must be on top of the stack in both cases. That is, `ret n` pops the RA and *then* pops `n` bytes.

So I don't know if it is actually slower or not, but it is not semantically equivalent, so the point is moot to some degree.

The actual alternative would be something like

    mov tmp, [rsp]
    add rsp, n
    mov [rsp], tmp
    ret

I kind of doubt that is any better than `ret n`, but it is certainly possible. Would be interesting to see what LLVM does here.

Reply:
Makes sense. I didn't know that `ret imm` would move the stack pointer after reading the return address.